### PR TITLE
Implement PromptSaveWindow and LoadSaveQuit game command

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
The LoadSaveQuit game command is intertwined with the PromptSaveWindow. I am implementing both in this PR.

### CODE REVIEW

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+        if (Input::hasFlag(Input::input_flags::flag5))
+        {
+            Input::sub_407231();
+            Input::resetFlag(Input::input_flags::flag5);
+        }
+
+        Title::start();
+
+        // TODO: interrupt tick?
```
Comment (AaronVanGeffen) [inline src/OpenLoco/GameCommands/LoadSaveQuit.cpp:71]: The original performs a `mov esp, game_loop_iteration_esp` here. Do we need to adapt this?

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    {
+        _game_command_nest_level = 0;
+
+        if (isNetworked())
+        {
+            clearScreenFlag(ScreenFlags::networked);
+        }
+
+        SDL_Quit();
+        exit(0);
```
Comment (AaronVanGeffen) [inline src/OpenLoco/GameCommands/LoadSaveQuit.cpp:40]: Temporary code; this should not go here.

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    {
+        if (isNetworked())
+        {
+            Ui::WindowManager::closeAllFloatingWindows();
+        }
+
+        Ui::WindowManager::close(Ui::WindowType::options);
+        Ui::WindowManager::close(Ui::WindowType::companyFaceSelection);
+        Ui::WindowManager::close(Ui::WindowType::objectSelection);
+        Ui::WindowManager::close(Ui::WindowType::saveGamePrompt);
```
Comment (AaronVanGeffen) [inline src/OpenLoco/GameCommands/LoadSaveQuit.cpp:57]: The original closes the save game prompt this close after it's been opened. How can the user ever interact with it?

**Comment by AaronVanGeffen** [comment]
The `LoadSaveQuit.cpp` is branching out quite widely, beyond what I feel is appropriate for a game command file. Perhaps they should be moved into a controller of some sort, like we have done for the editor controller? Perhaps in the `Ui` namespace? I would appreciate some suggestions.

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+                Ui::Windows::Error::open(StringIds::error_the_other_player_has_exited_the_game, StringIds::null);
+
+                throw GameException::Interrupt;
+            }
+        }
+
+        // 0x0043C1C8
+        exitCleanly();
+        return 0;
```
Comment (AaronVanGeffen) [inline src/OpenLoco/GameCommands/LoadSaveQuit.cpp:162]: The MSVC CI complains that this line is never reached — and it isn't, because `exitCleanly` exits the game.
```
  D:\a\OpenLoco\OpenLoco\src\OpenLoco\GameCommands\LoadSaveQuit.cpp(162): warning C4702: unreachable code [D:\a\OpenLoco\OpenLoco\src\OpenLoco\openloco.vcxproj]
```
However, if I leave it out, both gcc and clang complain…

Comment (duncanspumpkin) [reply src/OpenLoco/GameCommands/LoadSaveQuit.cpp:162]: Just add a pragma to ignore the warning

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+            return 0;
+
+        if (dl == 1)
+        {
+            Ui::WindowManager::close(Ui::WindowType::saveGamePrompt);
+            return 0;
+        }
+
+        if (dl == 2)
+            return 0;
```
Comment (duncanspumpkin) [inline src/OpenLoco/GameCommands/LoadSaveQuit.cpp:278]: In original this would go to 0x0043BFE3 why the change?

File: src/OpenLoco/Windows/PromptSaveWindow.cpp [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+            if (window == nullptr)
+                return nullptr;
+
+            window->widgets = _widgets;
+            window->enabled_widgets = (1 << widx::closeButton) | (1 << widx::saveButton) | (1 << widx::dontSaveButton) | (1 << widx::cancelButton);
+            window->initScrollWidgets();
+            window->colours[0] = Colour::translucent(Colour::salmon_pink);
+            window->flags |= Ui::WindowFlags::transparent;
+
+            setPauseFlag(2);
```
Comment (duncanspumpkin) [inline src/OpenLoco/Windows/PromptSaveWindow.cpp:65]: 1 << 1

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+
+        _savePromptType = di;
+        Ui::TextInput::cancel();
+        Ui::Windows::PromptSaveWindow::open(_savePromptType);
+
+        if (!isTitleMode())
+        {
+            // 0x0043C369
+            if (Tutorial::state() == Tutorial::tutorial_state::playing)
```
Comment (duncanspumpkin) [inline src/OpenLoco/GameCommands/LoadSaveQuit.cpp:287]: The tutotial_state::recording branch has been removed. Was that a conscious decision

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+            {
+                Tutorial::stop();
+            }
+            else if (!isNetworked() || _savePromptType != 2)
```
Comment (duncanspumpkin) [inline src/OpenLoco/GameCommands/LoadSaveQuit.cpp:291]: surely we know all of the save prompt types.

Comment (AaronVanGeffen) [reply src/OpenLoco/GameCommands/LoadSaveQuit.cpp:291]: Yeah, there's only three. I should add an enum for them.

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp (hunk 7 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/GameCommands/LoadSaveQuit.cpp:287]: It was, yeah. Previously, we talked about how the current tutorial system should be replaced with something that isn't pixel/coordinate based at some point. So I consider the current system obsolete.

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+        return regs.eax != 0;
+    }
+
+    // 0x0043BFF8
+    static uint32_t loadGame()
+    {
+        GameCommands::do_21(1, 0);
+        Input::toolCancel();
+
+        if (isEditorMode() && loadLandscapeOpen())
```
Comment (duncanspumpkin) [inline src/OpenLoco/GameCommands/LoadSaveQuit.cpp:84]: I'm not sure having the if like this is quite the same. If loadLandscapeOpen returns false this version will move to loadSaveGameOpen. This is different to vanilla where it would leave the if else chain.

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+                throw GameException::Interrupt;
+            }
+        }
+        else if (!isNetworked() && loadSaveGameOpen())
+        {
+            // 0x0043C033
+            auto path = fs::path(&_savePath[0]).replace_extension(S5::extensionSV5).u8string();
+            std::strncpy(&_currentScenarioFilename[0], path.c_str(), std::size(_currentScenarioFilename));
+
+            if (sub_441FA7())
```
Comment (duncanspumpkin) [inline src/OpenLoco/GameCommands/LoadSaveQuit.cpp:102]: I've this function labelled as `loadTitleSequence` in my IDA but don't know why.

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+
+            // If the other party is leaving the game, go back to the title screen.
+            if (playerCompanyId != previousUpdatingId)
+            {
+                // 0x0043C1CD
+                addr<0x00F25428, uint32_t>() = 0;
+                clearScreenFlag(ScreenFlags::networked);
+                clearScreenFlag(ScreenFlags::networkHost);
+                addr<0x00508F0C, uint32_t>() = 0;
+                addr<0x00525E3C, uint8_t>() = 0;
```
Comment (duncanspumpkin) [inline src/OpenLoco/GameCommands/LoadSaveQuit.cpp:149]: These are already known offsets playerCompanyId...

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+
+            Ui::WindowManager::closeAllFloatingWindows();
+
+            CompanyManager::updatingCompanyId(previousUpdatingId);
+            setScreenFlag(ScreenFlags::networked);
+
+            // If the other party is leaving the game, go back to the title screen.
+            if (playerCompanyId != previousUpdatingId)
+            {
+                // 0x0043C1CD
```
Comment (duncanspumpkin) [inline src/OpenLoco/GameCommands/LoadSaveQuit.cpp:144]: Do we know what the missing function call to `0x406B2F` was?

File: src/OpenLoco/OpenLoco.cpp [hunk 13]
```diff
@@ -329,6 +331,21 @@ namespace OpenLoco
         call(0x004BE5EB, regs);
     }
 
+    // 0x004BE65E
+    void exitCleanly()
+    {
+        auto tempFilePath = Environment::getPathNoWarning(Environment::path_id::_1tmp);
```
Comment (duncanspumpkin) [inline src/OpenLoco/OpenLoco.cpp:337]: What about the sending out of the quit messages and calling audio_close?

File: src/OpenLoco/Title.cpp [hunk 14]
```diff
@@ -62,7 +63,7 @@ namespace OpenLoco::Title
         CompanyManager::updatingCompanyId(CompanyManager::getControllingId());
         if (isPaused())
         {
-            togglePause(true);
+            GameCommands::do_20();
```
Comment (duncanspumpkin) [inline src/OpenLoco/Title.cpp:66]: Is this safe to do? Will the game command actually be executed?

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+
+        throw GameException::Interrupt;
+    }
+
+    // 0x0043C427
+    void confirmSaveGame()
+    {
+        Input::toolCancel();
+
+        if (isEditorMode() && saveLandscapeOpen())
```
Comment (duncanspumpkin) [inline src/OpenLoco/GameCommands/LoadSaveQuit.cpp:215]: as above

**Review by duncanspumpkin — CHANGES_REQUESTED** [review]
See comments

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp (hunk 12 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/GameCommands/LoadSaveQuit.cpp:144]: No, to be honest. It seemed like a wrapper for Win32-related cleanup logic. I haven't spent much on it, other than that.

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp (hunk 10 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/GameCommands/LoadSaveQuit.cpp:102]: I have it labelled such over here as well, but I believe that's wrong. It's clearly more general than that, loading saved games in general.

File: src/OpenLoco/OpenLoco.cpp (hunk 13 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/OpenLoco.cpp:344]: As I understand it, `post_quit_message` is setting the return code. I think `exit(0)` covers that.

You're right about `audio_close` being omitted. We haven't implemented that function yet, though, so that needs doing as well.

File: src/OpenLoco/Title.cpp (hunk 14 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/Title.cpp:146]: Our old `togglePause` was incomplete compared to what the game command does. However, what I'm doing here is still different, as the original routine called `0x00431E32` (the togglePause game command) directly. Any thoughts?

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp (hunk 7 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/GameCommands/LoadSaveQuit.cpp:300]: I've added a comment indicating it has been omitted.

File: src/OpenLoco/GameCommands/LoadSaveQuit.cpp (hunk 5 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/GameCommands/LoadSaveQuit.cpp]: I think I made a mistake here. I've changed the logic to match vanilla.

File: src/OpenLoco/Title.cpp (hunk 14 above)
Comment (duncanspumpkin) [reply src/OpenLoco/Title.cpp:146]: I'm just slightly concerned that during loading/saving the game command runner might be in a state that disallows the toggle pause command. I don't know enough about the game command runner to be sure this isn't an issue.

Comment (AaronVanGeffen) [reply src/OpenLoco/Title.cpp]: Given the `togglePause` game command is rather short to begin with, let's just implement it to hopefully solve this potential problem.

**Comment by duncanspumpkin** [comment]
could you rebase this and then i think its ready to go

File: src/OpenLoco/OpenLoco.cpp [hunk 16]
```diff
@@ -336,6 +338,23 @@ namespace OpenLoco
         call(0x004BE5EB, regs);
     }
 
+    // 0x004BE65E
+    void exitCleanly()
```
Comment (ZehMatt) [inline src/OpenLoco/OpenLoco.cpp:342]:
```suggestion
    [[ noreturn ]] void exitCleanly()
```

Comment (ZehMatt) [reply src/OpenLoco/OpenLoco.cpp:342]: Your call to exit will never return. It can potentially cause compiler warnings if you rely on this.

**Comment by AaronVanGeffen** [comment]
From discussion on Discord, it seems moving functions might be best left to a near-future refactor involving involving `OpenLoco.cpp` as well.

I'll rebase this to account for #877 later tonight, and will clean up the commit history a little as well.

File: src/OpenLoco/OpenLoco.cpp (hunk 16 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/OpenLoco.cpp]: Right, we indeed had to deal with MSVC warning C4702. Your solution is better than disabling the warning — let's go with that.

**Review by duncanspumpkin — APPROVED** [review]
Assume you have tested this a good bit. I've only done a code review

File: src/OpenLoco/OpenLoco.cpp (hunk 16 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/OpenLoco.cpp]: Turns out it's still triggering the C4702 warning, even with the noreturn attribute. I will reinstate the pragma line suppressing it, then.

Comment (ZehMatt) [reply src/OpenLoco/OpenLoco.cpp]: The reason for noreturn would be something lets say something like this:
```
[[noreturn]] void exitCleanly();

int foo()
{
   if(_requestExit)
   {
       exitCleanly();
   }
   else
   {
       return 1;
   }
}
```
This would warn that not all paths return a value if exitCleanly would not have the ```[[noreturn]]``` attribute
